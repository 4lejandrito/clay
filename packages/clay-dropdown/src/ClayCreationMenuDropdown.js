import 'clay-icon';
import './ClayDropdownBase';
import Component from 'metal-component';
import defineWebComponent from 'metal-web-component';
import Soy from 'metal-soy';
import {Config} from 'metal-state';

import itemsValidator from './items_validator';
import templates from './ClayCreationMenuDropdown.soy.js';

/**
 * Implementation for Metal Clay Action Dropdown.
 */
class ClayCreationMenuDropdown extends Component {
	/**
	 * Handles footer button click.
	 * @param {!Event} event
	 * @protected
	 */
	_handleButtonClick(event) {
		this.emit('moreButtonClicked', event);
	}

	/**
	 * Continues the propagation of the item clicked event
	 * @param {!Object} item
	 * @protected
	 */
	_handleItemClick(item) {
		this.emit('itemClicked', item);
	}
}

/**
 * State definition.
 * @static
 * @type {!Object}
 */
ClayCreationMenuDropdown.STATE = {
	/**
	 * Button configuration to place a button at dropdown footer.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	button: Config.shapeOf({
		label: Config.string().required(),
		style: Config.oneOf(['primary', 'secondary']).value('primary'),
		type: Config.oneOf(['button', 'reset', 'submit']),
	}),

	/**
	 * Caption text of the dropdown.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	caption: Config.string(),

	/**
	 * Flag to indicate if menu is disabled
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?bool}
	 * @default false
	 */
	disabled: Config.bool().value(false),

	/**
	 * CSS classes to be applied to the element.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	elementClasses: Config.string(),

	/**
	 * Flag to indicate if menu is expanded.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?bool}
	 * @default false
	 */
	expanded: Config.bool().value(false),

	/**
	 * List of favorite menu items.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?Array|undefined}
	 * @default undefined
	 */
	favoriteItems: itemsValidator,

	/**
	 * Help text to be shown on top of the open dropdown.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	helpText: Config.string(),

	/**
	 * Id to be applied to the element.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	id: Config.string(),

	/**
	 * Position in which item icon will be placed. Needed if any item has icons.
	 * @instance
	 * @memberof ClayDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	itemsIconAlignment: Config.oneOf(['left', 'right']),

	/**
	 * List of main menu items.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {!Array}
	 * @default undefined
	 */
	mainItems: itemsValidator.required(),

	/**
	 * The path to the SVG spritemap file containing the icons.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {!string}
	 * @default undefined
	 */
	spritemap: Config.string().required(),

	/**
	 * CSS classes to be applied to the trigger element.
	 * @instance
	 * @memberof ClayCreationMenuDropdown
	 * @type {?string|undefined}
	 * @default undefined
	 */
	triggerClasses: Config.string(),
};

defineWebComponent('clay-creation-menu-dropdown', ClayCreationMenuDropdown);

Soy.register(ClayCreationMenuDropdown, templates);

export {ClayCreationMenuDropdown};
export default ClayCreationMenuDropdown;
