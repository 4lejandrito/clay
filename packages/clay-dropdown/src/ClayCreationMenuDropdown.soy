{namespace ClayCreationMenuDropdown}

/**
 * This renders an actions dropdown.
 */
{template .render}
	{@param mainItems: list<?>}
	{@param spritemap: string}
	{@param? _handleButtonClick: any}
	{@param? _handleItemClick: any}
	{@param? caption: string}
	{@param? disabled: bool}
	{@param? elementClasses: string}
	{@param? expanded: bool}
	{@param? favoriteItems: list<?>}
	{@param? helpText: string}
	{@param? id: string}
	{@param? itemsIconAlignment: string}
	{@param? triggerClasses: string}

  {let $maxFavoriteItems: 7 /}
  {let $maxMainItems: 8 /}
  {let $maxTotalItems: 15 /}

  {let $trigger kind="html"}
    {call ClayIcon.render}
      {param spritemap: $spritemap /}
      {param symbol: 'plus' /}
    {/call}
  {/let}

  {let $items: $favoriteItems ? [[
    'items': $mainItems,
    'maxItems': $maxMainItems,
    'separator': true,
    'type': 'group'
  ],[
    'items': $favoriteItems,
    'label': 'Favorites',
    'maxItems': $maxFavoriteItems,
    'type': 'group'
  ]] : [[
    'items': $mainItems,
    'maxItems': $maxTotalItems,
    'separator': false,
    'type': 'group'
  ]] /}

  {let $showMoreButton: (not $favoriteItems and $mainItems and length($mainItems) > $maxTotalItems) 
    or ($favoriteItems and length($favoriteItems) > $maxFavoriteItems)
    or ($mainItems and length($mainItems) > $maxMainItems) /}

	{call ClayDropdownBase.render}
    {param button: $showMoreButton ? [
      'label': 'More',
      'style': 'secondary'
    ] : null /}
    {param caption: $caption /}
    {param disabled: $disabled /}
    {param elementClasses: $elementClasses /}
    {param events: [
      'buttonClicked': $_handleButtonClick,
      'itemClicked': $_handleItemClick
    ]/}
    {param expanded: $expanded /}
    {param helpText: $helpText /}
    {param id: $id /}
    {param items: $items /}
    {param itemsIconAlignment: $itemsIconAlignment /}
    {param label: $trigger /}
    {param ref: 'dropdown' /}
    {param spritemap: $spritemap /}
    {param style: 'primary' /}
    {param triggerClasses: $triggerClasses /}
  {/call}
{/template}
