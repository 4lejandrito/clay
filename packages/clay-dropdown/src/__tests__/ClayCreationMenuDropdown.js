import ClayCreationMenuDropdown from '../ClayCreationMenuDropdown';

let clayCreationMenuDropdown;

let getItems = function(maxItems, name) {
	let items = [];

	for (let i=1; i <= maxItems; i++) {
		items.push({
			href: '#' + name + i,
			label: name + i,
		});
	}

	return items;
};

describe('ClayCreationMenuDropdown', function() {
	afterEach(() => {
		if (clayCreationMenuDropdown) {
			clayCreationMenuDropdown.dispose();
		}
	});

	it('should render default markdown', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a creation menu dropdown with classes', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			elementClasses: 'my-custom-class',
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a creation menu dropdown with id', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			id: 'myId',
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a disabled creation menu dropdown', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			disabled: true,
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render an expanded creation menu dropdown', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			expanded: true,
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a creation menu dropdown with favorite items', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			favoriteItems: getItems(2, 'favorite'),
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a creation menu dropdown with 7 favorite items and more button when more than 7 favorite items are passed', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			favoriteItems: getItems(8, 'favorite'),
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a creation menu dropdown with 8 main items and more button when more than 8 main items and at least one favorite item are passed', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			favoriteItems: getItems(2, 'favorite'),
			mainItems: getItems(10, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a creation menu dropdown with 15 main items and more button when more than 15 main items and at no favorite item are passed', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			mainItems: getItems(20, 'main'),
			spritemap: 'icons.svg',
		});

		expect(clayCreationMenuDropdown).toMatchSnapshot();
	});

	it('should render a creation menu dropdown and emit an event on more button click', () => {
		clayCreationMenuDropdown = new ClayCreationMenuDropdown({
			favoriteItems: getItems(8, 'favorite'),
			mainItems: getItems(2, 'main'),
			spritemap: 'icons.svg',
		});

		const spy = jest.spyOn(clayCreationMenuDropdown, 'emit');

		const element =
			clayCreationMenuDropdown.refs.dropdown.refs.portal.refs
				.dropdownButton.element;
		element.click();

		expect(spy).toHaveBeenCalled();
		expect(spy).toHaveBeenCalledWith(
			'moreButtonClicked',
			expect.any(Object)
		);
	});
});
