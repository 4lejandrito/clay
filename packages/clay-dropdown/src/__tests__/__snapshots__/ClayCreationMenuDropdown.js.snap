// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayCreationMenuDropdown should render a creation menu dropdown with 7 favorite items and more button when more than 7 favorite items are passed 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <form>
    <div class="inline-scroller">
      <ul class="list-unstyled">
        <li>
          <a class=" dropdown-item" href="#main1">main1</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main2">main2</a>
        </li>
        <li aria-hidden="true" class="dropdown-divider" role="presentation"></li>
        <li class="dropdown-subheader" role="presentation">Favorites</li>
        <li>
          <a class=" dropdown-item" href="#favorite1">favorite1</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#favorite2">favorite2</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#favorite3">favorite3</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#favorite4">favorite4</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#favorite5">favorite5</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#favorite6">favorite6</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#favorite7">favorite7</a>
        </li>
      </ul>
    </div>
    <div class="dropdown-caption">Showing 9 of 10 elements</div>
    <div class="dropdown-section">
      <button class="btn btn-block btn-secondary" type="button">More</button>
    </div>
  </form>
</div>
`;

exports[`ClayCreationMenuDropdown should render a creation menu dropdown with 8 main items and more button when more than 8 main items and at least one favorite item are passed 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <form>
    <div class="inline-scroller">
      <ul class="list-unstyled">
        <li>
          <a class=" dropdown-item" href="#main1">main1</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main2">main2</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main3">main3</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main4">main4</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main5">main5</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main6">main6</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main7">main7</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main8">main8</a>
        </li>
        <li aria-hidden="true" class="dropdown-divider" role="presentation"></li>
        <li class="dropdown-subheader" role="presentation">Favorites</li>
        <li>
          <a class=" dropdown-item" href="#favorite1">favorite1</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#favorite2">favorite2</a>
        </li>
      </ul>
    </div>
    <div class="dropdown-caption">Showing 10 of 12 elements</div>
    <div class="dropdown-section">
      <button class="btn btn-block btn-secondary" type="button">More</button>
    </div>
  </form>
</div>
`;

exports[`ClayCreationMenuDropdown should render a creation menu dropdown with 15 main items and more button when more than 15 main items and at no favorite item are passed 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <form>
    <div class="inline-scroller">
      <ul class="list-unstyled">
        <li>
          <a class=" dropdown-item" href="#main1">main1</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main2">main2</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main3">main3</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main4">main4</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main5">main5</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main6">main6</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main7">main7</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main8">main8</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main9">main9</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main10">main10</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main11">main11</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main12">main12</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main13">main13</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main14">main14</a>
        </li>
        <li>
          <a class=" dropdown-item" href="#main15">main15</a>
        </li>
      </ul>
    </div>
    <div class="dropdown-caption">Showing 15 of 20 elements</div>
    <div class="dropdown-section">
      <button class="btn btn-block btn-secondary" type="button">More</button>
    </div>
  </form>
</div>
`;

exports[`ClayCreationMenuDropdown should render a creation menu dropdown with classes 1`] = `
<div class="dropdown my-custom-class">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="#main1">main1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#main2">main2</a>
    </li>
  </ul>
</div>
`;

exports[`ClayCreationMenuDropdown should render a creation menu dropdown with favorite items 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="#main1">main1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#main2">main2</a>
    </li>
    <li aria-hidden="true" class="dropdown-divider" role="presentation"></li>
    <li class="dropdown-subheader" role="presentation">Favorites</li>
    <li>
      <a class=" dropdown-item" href="#favorite1">favorite1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#favorite2">favorite2</a>
    </li>
  </ul>
</div>
`;

exports[`ClayCreationMenuDropdown should render a creation menu dropdown with id 1`] = `
<div class="dropdown" id="myId">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="#main1">main1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#main2">main2</a>
    </li>
  </ul>
</div>
`;

exports[`ClayCreationMenuDropdown should render a disabled creation menu dropdown 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" disabled="" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="#main1">main1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#main2">main2</a>
    </li>
  </ul>
</div>
`;

exports[`ClayCreationMenuDropdown should render an expanded creation menu dropdown 1`] = `
<div class="dropdown show">
  <button aria-expanded="true" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu show" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="#main1">main1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#main2">main2</a>
    </li>
  </ul>
</div>
`;

exports[`ClayCreationMenuDropdown should render default markdown 1`] = `
<div class="dropdown">
  <button aria-expanded="false" aria-haspopup="true" aria-label="toggle" class="dropdown-toggle btn btn-primary" type="button">
    <svg aria-hidden="true" class="lexicon-icon lexicon-icon-plus" focusable="false">
      <title>plus</title>
      <use xlink:href="icons.svg#plus"></use>
    </svg>
  </button>
</div>
<div class="dropdown-menu" ref="menu">
  <ul class="list-unstyled">
    <li>
      <a class=" dropdown-item" href="#main1">main1</a>
    </li>
    <li>
      <a class=" dropdown-item" href="#main2">main2</a>
    </li>
  </ul>
</div>
`;
